### ISSUE TITLE
Race Condition in POST /1.0/storage-pools

### ISSUE THREAD

**Issue by ExternalReality** [lxc/lxd#4150]
# Required information

 * Distribution: ubuntu
 * Distribution version: 17.10
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.13.0-21-generic
   * LXC version: 2.18
   * LXD version: 2.21
   * Storage backend in use: zfs/dir/btrfs

# Issue description

When issuing multiple `POST 1.0/storage-pool` requests in parallel some of the requests succeed, while others fail with one of two different errors:

1. {"error":"The storage pool already exists","error_code":500,"type":"error"}
2. {"error":"Failed to create the ZFS pool: cannot create 'juju-zfs': pool already exists\n","error_code":500,"type":"error"} 

From looking at the code, in general, processes are creating and deleting the sqlite storage pool database record and zfs pool both due to processes entering critical sections that unwind work on exit. For example, if one process fails to create the zfs pool since it already exists(e.g. another processed just created it), [then this will cause the sqlite database record to be undone](https://github.com/lxc/lxd/blob/30442c88580d18f2703862f414c3cfba713caf13/lxd/storage_pools.go#L106-L125) which allows more processes to enter the critical sections of the storage-pool/create handler.

# Steps to reproduce

 1. parallel -j --lb 'lxdrace.sh' ::: \`seq 10\`

where lxdrace.sh is:

```sh
curl -s -H "Content-Type: application/json" -X POST -d '{ "config": { "size": "15GB" }, "driver": "zfs", "name": "some-pool" }' --unix-socket $LXD_DIR/unix.socket s/1.0/storage-pools
```

**Comment by stgraber** [comment]
What's your use case for attempting to create the same storage pools several times in parallel?
How did you end up in this situation?

There are quite a few places in LXD where you're effectively racing with the kernel or with other parts of the operating system. We could try to setup locks all over the place to prevent those from occurring but locks are costly and easy to forget or get wrong, causing much more annoying problems later down the road.

We've so far been doing work on those issues only when a legitimate workload would hit such issues or when the resulting state from hitting such an issue would lead to a broken (inconsistent) system state.
I don't believe this issue qualifies so I'm wary of adding even more complexity to the storage pool creation code to effectively slow things down so that you get a consistent error.

**Comment by stgraber** [comment]
Any fix for this would effectively end up guaranteeing that one of the requests will succeed (assuming no other problems) and that the other 9 will fail once that first one succeeds, but which will be the one succeeding would still be random based on exactly what goroutine is the first to acquire the lock.

So I'm not sure how this would really help a higher level tool like Juju since it'd still mostly be getting error back (as it should), I think the real question is why is it attempting to create the same storage pool more than once :)

**Comment by stgraber** [comment]
A pretty cheap fix for us is to just serialize the calls to POST for /1.0/storage-pools and /1.0/networks (that would likely suffer from the same issue), since those would just be two simple locks in two simple functions, it's pretty unlikely that mistakes would happen with them (nothing other than an http query will cause a call to those two functions).

**Comment by jameinel** [comment]
So I'm not sure the issue is a serialization of POST to storage-pools. The issue for us is not that one works and others fail (as that is expected).
Our function is an "ensure this storage pool exists with the right parameters".
We could arguably restructure it to lookup the pool first, and only if it gets a NotFound fall back to another error.
The issue that I ran into during testing, is that if the *ZFS* pool already exists, but the LXD agent doesn't know about it, then CreateStoragePool fails with "pool already exists", but *sometimes* StoragePoolGet fails with NotFound (and sometimes it doesn't fail).
When I encountered it, it was because I had switched from LXD from debian, to LXD from snap, both using the same filesystem (presumably). So I had a "juju-zfs" pool already, but the LXD I was connecting to didn't know about it.
I don't know exactly why "StoragePoolGet" will succeed if only the zfs pool exists, but *not* a database record.
(maybe I would have expected Get to fail, but that Create would succeed and just record the database entry for the existing pool.)

The issue is that Create isn't doing anything (it always gets a "pool already exists" error), but sometimes while Create is failing, we also have StoragePoolGet return a valid pool, and sometimes it returns a NotFound error.
